symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java

```diff
@@ -67,31 +92,120 @@ public void start(Stage stage) {
+
+      // create a service feature table
+      ServiceFeatureTable serviceFeatureTable = new ServiceFeatureTable("https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/ArcGIS/rest/services/Redlands_Restaurants/FeatureServer/0");
+      // create the restaurants feature layer using the service feature table
+      FeatureLayer restaurantLayer = new FeatureLayer(serviceFeatureTable);
+
+      // map the input fields in the feature layer to the dictionary symbol style's expected fields for symbols and text
+      Map<String, String> fieldMap = new HashMap<>();
+      fieldMap.put("healthgrade", "Inspection");
+
+      toggleGroup.selectedToggleProperty().addListener(changed -> {
+        // clear the existing dictionary symbol style and feature layer on the map
+        map.getOperationalLayers().clear();
+        dictionarySymbolStyle = null;
+        dictionaryRenderer = null;
+
+        // turn on the progress indicator when the symbols are loading
+        progressIndicator.setVisible(true);
+
+        if (toggleGroup.getSelectedToggle() == webStyleButton) {
+          dictionarySymbolStyle = new DictionarySymbolStyle(portalItem);
+          // load the symbol dictionary
+          dictionarySymbolStyle.loadAsync();
+          dictionarySymbolStyle.addDoneLoadingListener(() -> {
+            if (dictionarySymbolStyle.getLoadStatus() == LoadStatus.LOADED) {
+              // create a dictionary renderer with the dictionary symbol style
+              dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle, fieldMap, fieldMap);
+              // renderer tells layer what symbols to apply to what features
+              restaurantLayer.setRenderer(dictionaryRenderer);
+              // add the layer to the map
+              map.getOperationalLayers().add(restaurantLayer);
+              // hide the progress indicator when the symbols have been applied to the features
+              progressIndicator.setVisible(false);
+            } else {
+              new Alert(Alert.AlertType.ERROR, "Dictionary symbol style failed to load!").show();
+            }
+          });
+        }
+        if (toggleGroup.getSelectedToggle() == fileStyleButton){
+          // create a symbol dictionary from the local symbol style file
+          dictionarySymbolStyle = DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath());
+          // create a dictionary renderer with the dictionary symbol style
+          dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle);
+          // renderer tells layer what symbols to apply to what features
+          restaurantLayer.setRenderer(dictionaryRenderer);
+          // add the layer to the map
+          map.getOperationalLayers().add(restaurantLayer);
+          // hide the progress indicator when the symbols have been applied to the features
+          progressIndicator.setVisible(false);
+        }
+      });
```

Comment:
I'm sure you've thought of this one before: I originally saw there might be a bit of scope for refactoring these three lines here, by adding them to the end of the statement `toggleGroup.selectedToggleProperty().addListener(changed -> {`
It would mean that the statements `if (toggleGroup.getSelectedToggle() == webStyleButton/fileStyleButton` would only be responsible for populating the `dictionaryRenderer`, and after that, the renderer/map/progress indicator get dealt with either way.

However, the whoe async process under `if (toggleGroup.getSelectedToggle() == webStyleButton) {` kind of complicates the whole thing! 
Can't think of a different pattern just now, and I don't think it's worth trying to come up with something less readable than what we've got at the moment.

```suggestion
      toggleGroup.selectedToggleProperty().addListener(changed -> {
        // clear the existing dictionary symbol style and feature layer on the map
        map.getOperationalLayers().clear();
        dictionarySymbolStyle = null;
        dictionaryRenderer = null;

        // turn on the progress indicator when the symbols are loading
        progressIndicator.setVisible(true);

		// create the renderer based on which source is selected for the symbol style
        if (toggleGroup.getSelectedToggle() == webStyleButton) {
          dictionarySymbolStyle = new DictionarySymbolStyle(portalItem);
          // load the symbol dictionary
          dictionarySymbolStyle.loadAsync();
          dictionarySymbolStyle.addDoneLoadingListener(() -> {
            if (dictionarySymbolStyle.getLoadStatus() == LoadStatus.LOADED) {
              // create a dictionary renderer with the dictionary symbol style
              dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle, fieldMap, fieldMap);
            } else {
              new Alert(Alert.AlertType.ERROR, "Dictionary symbol style failed to load!").show();
            }
          });
        }
        if (toggleGroup.getSelectedToggle() == fileStyleButton){
          // create a symbol dictionary from the local symbol style file
          dictionarySymbolStyle = DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath());
          // create a dictionary renderer with the dictionary symbol style
          dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle);
        }
        
        // set the renderer to the feature layer
        restaurantLayer.setRenderer(dictionaryRenderer);
        // add the layer to the map
        map.getOperationalLayers().add(restaurantLayer);
        // hide the progress indicator when the symbols have been applied to the features
        progressIndicator.setVisible(false);
      });
```